Port to jbuilder

mseri opened the pull request:
@jonludlam was the library unwrapped/unpackaged on purpose?

Commit:
Port to jbuilder

The libraries where not packed, so
although I believe the wrapping into
namespace is the correct thing to do
I will add a commit to unwrap them
for retrocompatibility

Signed-off-by: Marcello Seri <marcello.seri@citrix.com>

Commit:
Unwrap the library for retrocompatibility

Signed-off-by: Marcello Seri <marcello.seri@citrix.com>

mseri commented:
@gaborigloi can you review this? (plus it does seem we do not have merge rights)

gaborigloi commented:
@mseri sure :)

gaborigloi commented:
Yep, we don't have too much rights here - I can't even assign myself as a reviewer XD

gaborigloi commented:
@mseri There are warnings that we could try removing, to make `make` work with all the warnings-as-errors enabled.

gaborigloi commented on Makefile:
```diff
@@ -1,44 +1,28 @@
+reindent:
+	ocp-indent --syntax cstruct -i lib/*.mli
+	ocp-indent --syntax cstruct -i lib/*.ml
+	ocp-indent --syntax cstruct -i lib_test/*.ml
```
I like these new `make reindent` targets, will make it easier for contributors to indent.

gaborigloi commented on ezxenstore.opam:
```diff
@@ -6,19 +6,13 @@ homepage: "https://github.com/xapi-project/ezxenstore"
 depends: [
+  "jbuilder" {build}
+  "cmdliner"
+  "logs"
```
I really like the alphabetical ordering, much better than random :smile:

mseri commented:
I was planning on removing them, but I believe they could be breaking changes...

The warnings are about unused `~xs` variables in `with_xs` (https://github.com/xapi-project/ezxenstore/blob/master/lib/watch.ml#L27) and `wait_for` https://github.com/xapi-project/ezxenstore/blob/master/lib/watch.ml#L31). Maybe they are there to keep the interface retro-compatible? ping @jonludlam

mseri commented:
If we are to release a new major version, I think we should remove the `(wrapped false)` and clean up the signature of the `watch` functions. So we should probably postpone that to the future.

gaborigloi commented on ezxenstore.opam:
```diff
@@ -6,19 +6,13 @@ homepage: "https://github.com/xapi-project/ezxenstore"
-  "ocamlfind" "remove" "ezxenstore"
-]
+build: [[ "jbuilder" "build" "-p" name "-j" jobs ]]
+build-test: [[ "jbuilder" "runtest" "-p" name "-j" jobs ]]
```
Maybe we should not add this, because these are not unit tests at the moment - this requires a running xenstored to work, therefore the tests fail on my machine.

Commit:
opam: remove build-test

The tests are not unit tests but need to be run on a machine with `xenstored`.
They can be run with `jbuilder runtest` or `make test`

Signed-off-by: Marcello Seri <marcello.seri@citrix.com>

gaborigloi commented on lib/jbuild:
```diff
@@ -0,0 +1,11 @@
+ ((name ezxenstore)
+  (public_name ezxenstore)
+  (modules (xenstore watch))
+  (wrapped false)
```
Ah I see, so this is for compatibility, to not break code that uses this library by putting everything under one module that previously was top-level.

gaborigloi replied:
Jbuilder's documentation nicely explained it to me: https://jbuilder.readthedocs.io/en/latest/jbuild.html#library

mseri replied:
Yep, it's making namespacing (kind of) the default :tada: 
I think it should be always `true` (the default) if possible

gaborigloi commented on Makefile:
```diff
@@ -1,44 +1,28 @@
-
-# OASIS_STOP
+reindent:
+	ocp-indent --syntax cstruct -i lib/*.mli
```
Really minor issue, but `make reindent` fails due to the lack of `mli`s in `lib/`:
```
> make reindent
ocp-indent --syntax cstruct -i lib/*.mli
ocp-indent: FILE... arguments: no `lib/*.mli' file or directory
Usage: ocp-indent [OPTION]... [FILE]...
Try `ocp-indent --help' for more information.
Makefile:26: recipe for target 'reindent' failed
make: *** [reindent] Error 1
```

gaborigloi replied:
Maybe we could remove this line for now to fix the issue.

gaborigloi commented on Makefile:
```diff
@@ -1,44 +1,28 @@
-
-# OASIS_STOP
+reindent:
+	ocp-indent --syntax cstruct -i lib/*.mli
```
We could say instead something like:
```
ocp-indent --syntax cstruct -i lib/*.ml lib/*.mli
```

Commit:
Fix `make reindent` and use it on the sources

Signed-off-by: Marcello Seri <marcello.seri@citrix.com>

Commit:
Makefile: make 'make reindent' more general

Signed-off-by: Marcello Seri <marcello.seri@citrix.com>

gaborigloi reviewed (approved):
Looks great, feel free to squash & merge (if you'd like to, it's not necessary in my opinion)
